CDI will not create an EJBDescriptor for archive containing bean-discovery-mode=none.

**Describe the bug**  
CDI will not create an EJBDescriptor for archive containing bean-discovery-mode=none. However MDBs are not CDI beans, this means that CDI will fail to create a managed object for MDBs in an archive with bean-discovery-mode=none even though it should.

Stack trace:

`java.lang.NullPointerException: Cannot invoke "org.jboss.weld.ejb.spi.EjbDescriptor.isMessageDriven()" because "ejbDescriptor" is null at com.ibm.ejs.container.util.ExceptionUtil.EJBException(ExceptionUtil.java:401) at com.ibm.ejs.container.util.ExceptionUtil.EJBException(ExceptionUtil.java:287) at com.ibm.ejs.container.BusinessExceptionMappingStrategy.mapException(BusinessExceptionMappingStrategy.java:349) at com.ibm.ejs.container.BusinessExceptionMappingStrategy.setUncheckedException(BusinessExceptionMappingStrategy.java:521) at com.ibm.ejs.container.EJSDeployedSupport.setUncheckedException(EJSDeployedSupport.java:421) at com.ibm.ejs.container.EJSContainer.preinvokeHandleException(EJSContainer.java:2654) at com.ibm.ejs.container.EJSContainer.EjbPreInvoke(EJSContainer.java:2435) at com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.EJSLocal0SLUndiscoveredStatelessBean_8678189b.test(EJSLocal0SLUndiscoveredStatelessBean_8678189b.java) at com.ibm.ws.cdi.ejb.apps.ejbdiscovery.servlet.DiscoveryServlet.testCallNotDiscoveredBean(DiscoveryServlet.java:111) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:568) at componenttest.app.FATServlet.doGet(FATServlet.java:73) at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:527) at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:614) at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1260) at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:748) at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:445) at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1361) at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1077) at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:77) at com.ibm.ws.webcontainer40.servlet.CacheServletWrapper40.handleRequest(CacheServletWrapper40.java:87) at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:969) at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:293) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1246) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:468) at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:427) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:569) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:503) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:363) at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:330) at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:169) at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:77) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:516) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:586) at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:970) at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1059) at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:247) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) at java.base/java.lang.Thread.run(Thread.java:858) Caused by: java.lang.NullPointerException: Cannot invoke "org.jboss.weld.ejb.spi.EjbDescriptor.isMessageDriven()" because "ejbDescriptor" is null at com.ibm.ws.cdi.impl.managedobject.CDIEJBManagedObjectFactoryImpl.createContext(CDIEJBManagedObjectFactoryImpl.java:84) at com.ibm.ejs.container.ManagedBeanOBase.createInterceptorsAndInstance(ManagedBeanOBase.java:200) at com.ibm.ejs.container.StatelessBeanO.initialize(StatelessBeanO.java:157) at com.ibm.ejs.container.BeanOFactory.create(BeanOFactory.java:105) at com.ibm.ejs.container.EJSHome.createBeanO(EJSHome.java:932) at com.ibm.ejs.container.EJSHome.createBeanO(EJSHome.java:1029) at com.ibm.ejs.container.activator.UncachedActivationStrategy.atActivate(UncachedActivationStrategy.java:64) at com.ibm.ejs.container.activator.Activator.preInvokeActivateBean(Activator.java:265) at com.ibm.ejs.container.EJSContainer.preInvokeActivate(EJSContainer.java:3003) at com.ibm.ejs.container.EJSContainer.EjbPreInvoke(EJSContainer.java:2392) `

**Steps to Reproduce**  
Create an MDB inside an archive with bean-discovery-mode=none 

**Expected behavior**  
The MDB should run despite bean-discovery-mode=none 

**Diagnostic information:**  
 - OpenLiberty Version: 18.0.0.2+
 - Affected feature(s) [e.g. cdi-1.2+]

--- a/dev/com.ibm.ws.cdi.ejb_fat/bnd.bnd
+++ b/dev/com.ibm.ws.cdi.ejb_fat/bnd.bnd
@@ -43,6 +43,10 @@ tested.features:\
         xmlbinding-3.0,\
         xmlbinding-4.0,\
         enterprisebeanspersistenttimer-4.0,\
+        messaging-3.0,\
+        messaging-3.1,\
+        messagingServer-3.0,\
+        messagingClient-3.0,\
         expressionlanguage-6.0,\
         faces-5.0,\
         noship-1.0,\

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/ejbs/interfaces/StatelessLocal.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/ejbs/interfaces/StatelessLocal.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/none/UndiscoveredStatelessBean.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/none/UndiscoveredStatelessBean.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -14,12 +14,12 @@ package com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none;
 
 import javax.ejb.Stateless;
 
-import com.ibm.ws.cdi.ejb.apps.ejbdiscovery.ejbs.interfaces.StatelessLocal;
-
 /**
  * Session Bean implementation class StatelessBean
  */
 @Stateless
-public class UndiscoveredStatelessBean implements StatelessLocal {
+public class UndiscoveredStatelessBean implements UndiscoveredStatelessLocal {
+
+    public void test() {}
 
 }

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/none/UndiscoveredStatelessLocal.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/none/UndiscoveredStatelessLocal.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -15,6 +15,8 @@ package com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none;
 import javax.ejb.Local;
 
 @Local
-public interface StatelessLocal {
+public interface UndiscoveredStatelessLocal {
+
+    public void test();
 
 }

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/servlet/DiscoveryServlet.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/ejbdiscovery/servlet/DiscoveryServlet.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -17,12 +17,14 @@ import static org.junit.Assert.assertTrue;
 import java.lang.reflect.Type;
 import java.util.Set;
 
+import javax.ejb.EJB;
 import javax.inject.Inject;
 import javax.servlet.annotation.WebServlet;
 
 import org.junit.Test;
 
 import com.ibm.ws.cdi.ejb.apps.ejbdiscovery.extension.DiscoveryExtension;
+import com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.UndiscoveredStatelessLocal;
 
 import componenttest.app.FATServlet;
 
@@ -32,6 +34,9 @@ public class DiscoveryServlet extends FATServlet {
 
     @Inject
     private DiscoveryExtension extension;
+
+    @EJB
+    private UndiscoveredStatelessLocal undiscoveredBean;
 
     private static void assertContains(Set<?> set, Object contains) {
         assertTrue(contains + " not found in " + set, set.contains(contains));
@@ -99,6 +104,11 @@ public class DiscoveryServlet extends FATServlet {
         Set<Class<?>> beans = extension.getObservedBeans();
         // There is a stateless bean that should not be discovered because the .jar has discovery-mode=none
         assertNotContains(beans, com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.UndiscoveredStatelessBean.class);
+    }
+
+    @Test
+    public void testCallNotDiscoveredBean() throws Exception {
+        undiscoveredBean.test();
     }
 
 }

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jar/JarMdb.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jar/JarMdb.java
@@ -1,0 +1,41 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbTestEar.jar;
+
+import javax.ejb.MessageDriven;
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageListener;
+
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.lib.EarTestMessageHolder;
+
+@MessageDriven
+public class JarMdb implements MessageListener {
+
+    @Inject
+    private EarTestMessageHolder messageHolder;
+
+    /** {@inheritDoc} */
+    @Override
+    public void onMessage(Message message) {
+
+        System.out.println("JarMdb Message received: " + message);
+
+        try {
+            String body = message.getBody(String.class);
+            messageHolder.addMessage(body);
+        } catch (JMSException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jarNoBeanDiscovery/JarMdbNotDiscovered.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jarNoBeanDiscovery/JarMdbNotDiscovered.java
@@ -1,0 +1,41 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbTestEar.jarNoBeanDiscovery;
+
+import javax.ejb.MessageDriven;
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageListener;
+
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.lib.EarTestMessageHolder;
+
+@MessageDriven
+public class JarMdbNotDiscovered implements MessageListener {
+
+    @Inject
+    private EarTestMessageHolder messageHolder;
+
+    /** {@inheritDoc} */
+    @Override
+    public void onMessage(Message message) {
+
+        System.out.println("JarMdbNotDiscovered Message received: " + message);
+
+        try {
+            String body = message.getBody(String.class);
+            messageHolder.addMessage(body);
+        } catch (JMSException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jarNoBeanDiscovery/beans.xml
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/jarNoBeanDiscovery/beans.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://java.sun.com/xml/ns/javaee"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+        xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
+        version="1.1"
+        bean-discovery-mode="none">
+       
+</beans>

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/lib/EarTestMessageHolder.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/lib/EarTestMessageHolder.java
@@ -1,0 +1,61 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbTestEar.lib;
+
+import static org.junit.Assert.fail;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class EarTestMessageHolder {
+
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
+    private static final Duration MIN_WAIT = Duration.ofMillis(10);
+
+    private final List<String> messages = new ArrayList<>();
+
+    /**
+     * Add a message to the holder
+     *
+     * @param message
+     */
+    public void addMessage(String message) {
+        synchronized (messages) {
+            messages.add(message);
+            messages.notifyAll();
+        }
+    }
+
+    /**
+     * Assert that a particular message is added within the timeout
+     *
+     * @param expectedMessage the message to wait for
+     * @throws InterruptedException if interrupted
+     */
+    public void assertMessageArrives(String expectedMessage) throws InterruptedException {
+        long startTime = System.nanoTime();
+        synchronized (messages) {
+            while (!messages.contains(expectedMessage)) {
+                Duration elapsed = Duration.ofNanos(System.nanoTime() - startTime);
+                Duration remaining = TIMEOUT.minus(elapsed);
+                if (remaining.minus(MIN_WAIT).isNegative()) {
+                    System.out.println("Giving up waiting");
+                    fail("Timed out waiting for message: " + expectedMessage + ", messages received:" + messages);
+                }
+                System.out.println("Still waiting for " + remaining);
+                messages.wait(remaining.toMillis());
+            }
+        }
+    }
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/war/EarMdbTestServlet.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbTestEar/war/EarMdbTestServlet.java
@@ -1,0 +1,73 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbTestEar.war;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Queue;
+import javax.jms.QueueConnection;
+import javax.jms.QueueConnectionFactory;
+import javax.jms.QueueSender;
+import javax.jms.QueueSession;
+import javax.jms.TextMessage;
+import javax.servlet.annotation.WebServlet;
+
+import org.junit.Test;
+
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.lib.EarTestMessageHolder;
+
+import componenttest.app.FATServlet;
+
+@WebServlet("/EarMdbTestServlet")
+public class EarMdbTestServlet extends FATServlet {
+
+    @Resource(name = "jms/qcf")
+    QueueConnectionFactory qcf;
+    @Resource(name = "jms/jar/jarQueue")
+    Queue jarQueue;
+    @Resource(name = "jms/jar/jarNoDiscoveryQueue")
+    Queue jarNoDiscoveryQueue;
+
+    @Inject
+    private EarTestMessageHolder messageHolder;
+
+    private static final long serialVersionUID = 1L;
+
+    @Test
+    public void testEarMdb() throws JMSException, InterruptedException {
+
+        try (QueueConnection conn = qcf.createQueueConnection()) {
+            conn.start();
+
+            QueueSession session = conn.createQueueSession(false, javax.jms.Session.AUTO_ACKNOWLEDGE);
+            QueueSender sender = session.createSender(jarQueue);
+
+            TextMessage message = session.createTextMessage();
+            message.setText("testJarMdb");
+            sender.send(message);
+            System.out.println("Servlet sent message: " + message);
+
+            messageHolder.assertMessageArrives("testJarMdb");
+
+            QueueSession jarNoDiscoverySession = conn.createQueueSession(false, javax.jms.Session.AUTO_ACKNOWLEDGE);
+            QueueSender jarNoDiscoverySender = jarNoDiscoverySession.createSender(jarNoDiscoveryQueue);
+
+            TextMessage jarNoDiscoveryMessage = session.createTextMessage();
+            jarNoDiscoveryMessage.setText("testJarMdbNoDiscovery");
+            jarNoDiscoverySender.send(jarNoDiscoveryMessage);
+            System.out.println("Servlet sent message: " + jarNoDiscoveryMessage);
+
+            messageHolder.assertMessageArrives("testJarMdbNoDiscovery");
+        }
+
+    }
+
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/BasicMdb.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/BasicMdb.java
@@ -1,0 +1,39 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbWar;
+
+import javax.ejb.MessageDriven;
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageListener;
+
+@MessageDriven
+public class BasicMdb implements MessageListener {
+
+    @Inject
+    private TestMessageHolder messageHolder;
+
+    /** {@inheritDoc} */
+    @Override
+    public void onMessage(Message message) {
+
+        System.out.println("Message received: " + message);
+
+        try {
+            String body = message.getBody(String.class);
+            messageHolder.addMessage(body);
+        } catch (JMSException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/BasicMdbTestServlet.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/BasicMdbTestServlet.java
@@ -1,0 +1,58 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbWar;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Queue;
+import javax.jms.QueueConnection;
+import javax.jms.QueueConnectionFactory;
+import javax.jms.QueueSender;
+import javax.jms.QueueSession;
+import javax.jms.TextMessage;
+import javax.servlet.annotation.WebServlet;
+
+import org.junit.Test;
+
+import componenttest.app.FATServlet;
+
+@WebServlet("/BasicMdbTestServlet")
+public class BasicMdbTestServlet extends FATServlet {
+
+    @Resource(name = "jms/qcf")
+    QueueConnectionFactory qcf;
+    @Resource(name = "jms/basic/queue")
+    Queue queue;
+
+    @Inject
+    private TestMessageHolder messageHolder;
+
+    private static final long serialVersionUID = 1L;
+
+    @Test
+    public void testBasicMdb() throws JMSException, InterruptedException {
+
+        try (QueueConnection conn = qcf.createQueueConnection()) {
+            conn.start();
+
+            QueueSession session = conn.createQueueSession(false, javax.jms.Session.AUTO_ACKNOWLEDGE);
+            QueueSender sender = session.createSender(queue);
+
+            TextMessage message = session.createTextMessage();
+            message.setText("testMdb");
+            sender.send(message);
+            System.out.println("Servlet sent message: " + message);
+        }
+
+        messageHolder.assertMessageArrives("testMdb");
+    }
+
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/TestMessageHolder.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/apps/mdbWar/TestMessageHolder.java
@@ -1,0 +1,61 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.apps.mdbWar;
+
+import static org.junit.Assert.fail;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class TestMessageHolder {
+
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
+    private static final Duration MIN_WAIT = Duration.ofMillis(10);
+
+    private final List<String> messages = new ArrayList<>();
+
+    /**
+     * Add a message to the holder
+     *
+     * @param message
+     */
+    public void addMessage(String message) {
+        synchronized (messages) {
+            messages.add(message);
+            messages.notifyAll();
+        }
+    }
+
+    /**
+     * Assert that a particular message is added within the timeout
+     *
+     * @param expectedMessage the message to wait for
+     * @throws InterruptedException if interrupted
+     */
+    public void assertMessageArrives(String expectedMessage) throws InterruptedException {
+        long startTime = System.nanoTime();
+        synchronized (messages) {
+            while (!messages.contains(expectedMessage)) {
+                Duration elapsed = Duration.ofNanos(System.nanoTime() - startTime);
+                Duration remaining = TIMEOUT.minus(elapsed);
+                if (remaining.minus(MIN_WAIT).isNegative()) {
+                    System.out.println("Giving up waiting");
+                    fail("Timed out waiting for message: " + expectedMessage + ", messages received:" + messages);
+                }
+                System.out.println("Still waiting for " + remaining);
+                messages.wait(remaining.toMillis());
+            }
+        }
+    }
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbDiscoveryTest.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/EjbDiscoveryTest.java
@@ -63,7 +63,7 @@ public class EjbDiscoveryTest extends FATServletClient {
         JavaArchive ejbDiscoveryNone = ShrinkWrap.create(JavaArchive.class,
                                                          EJB_DISCOVERY_APP_NAME + "None.jar")
                                                  .addClass(com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.UndiscoveredStatelessBean.class)
-                                                 .addClass(com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.StatelessLocal.class)
+                                                 .addClass(com.ibm.ws.cdi.ejb.apps.ejbdiscovery.none.UndiscoveredStatelessLocal.class)
                                                  .add(new FileAsset(new File("test-applications/" + EJB_DISCOVERY_APP_NAME + "None.jar/resources/WEB-INF/beans.xml")),
                                                       "/META-INF/beans.xml");
 

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/FATSuite.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/FATSuite.java
@@ -36,6 +36,7 @@ import componenttest.rules.repeater.RepeatTests;
                 EjbTimerTest.class,
                 InjectParameterTest.class,
                 InterceptorsTest.class,
+                MdbTest.class,
                 MultipleNamedEJBTest.class,
                 StatefulSessionBeanInjectionTest.class,
 })

--- a/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/MdbTest.java
+++ b/dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/MdbTest.java
@@ -1,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2023 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.cdi.ejb.tests;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.EnterpriseArchive;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.jar.JarMdb;
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.jarNoBeanDiscovery.JarMdbNotDiscovered;
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.lib.EarTestMessageHolder;
+import com.ibm.ws.cdi.ejb.apps.mdbTestEar.war.EarMdbTestServlet;
+import com.ibm.ws.cdi.ejb.apps.mdbWar.BasicMdbTestServlet;
+
+import componenttest.annotation.Server;
+import componenttest.annotation.TestServlet;
+import componenttest.annotation.TestServlets;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.rules.repeater.EERepeatActions;
+import componenttest.rules.repeater.RepeatTests;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+
+@RunWith(FATRunner.class)
+public class MdbTest extends FATServletClient {
+
+    public static final String SERVER_NAME = "cdi12MdbServer";
+    public static final String BASIC_APP = "basicMdb";
+    public static final String EAR_APP = "MdbTest";
+    public static final String EAR_WAR_NAME = "EarTestWar";
+
+    @ClassRule
+    public static RepeatTests r = EERepeatActions.repeat(SERVER_NAME, EERepeatActions.EE10, EERepeatActions.EE9, EERepeatActions.EE7);
+
+    @Server(SERVER_NAME)
+    @TestServlets({ @TestServlet(contextRoot = BASIC_APP, servlet = BasicMdbTestServlet.class),
+                    @TestServlet(contextRoot = EAR_WAR_NAME, servlet = EarMdbTestServlet.class)
+    })
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        WebArchive war = ShrinkWrap.create(WebArchive.class, BASIC_APP + ".war")
+                                   .addPackage(BasicMdbTestServlet.class.getPackage());
+
+        ShrinkHelper.exportDropinAppToServer(server, war, DeployOptions.SERVER_ONLY);
+
+        // Regular jar with MDB
+        JavaArchive ejbJar = ShrinkWrap.create(JavaArchive.class, "EjbJar.jar")
+                                       .addClass(JarMdb.class);
+
+        // Jar with MDB with bean-discovery-mode=none
+        JavaArchive ejbNoDiscoveryJar = ShrinkWrap.create(JavaArchive.class, "EjbNoDiscoveryJar.jar")
+                                                  .addClass(JarMdbNotDiscovered.class)
+                                                  .addAsManifestResource(JarMdbNotDiscovered.class.getResource("beans.xml"), "beans.xml");
+
+        // Ear lib, containing a CDI bean
+        JavaArchive earLib = ShrinkWrap.create(JavaArchive.class, "EarLib.jar")
+                                       .addClass(EarTestMessageHolder.class);
+
+        // War to hold the test servlet
+        WebArchive earTestWar = ShrinkWrap.create(WebArchive.class, EAR_WAR_NAME + ".war")
+                                          .addClass(EarMdbTestServlet.class);
+
+        EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, EAR_APP + ".ear")
+                                          .addAsModules(ejbJar, ejbNoDiscoveryJar, earTestWar)
+                                          .addAsLibrary(earLib);
+
+        ShrinkHelper.exportDropinAppToServer(server, ear, DeployOptions.SERVER_ONLY);
+
+        server.startServer();
+    }
+
+    @AfterClass
+    public static void teardown() throws Exception {
+        server.stopServer();
+    }
+}

--- a/dev/com.ibm.ws.cdi.ejb_fat/publish/servers/cdi12MdbServer/bootstrap.properties
+++ b/dev/com.ibm.ws.cdi.ejb_fat/publish/servers/cdi12MdbServer/bootstrap.properties
@@ -1,0 +1,2 @@
+bootstrap.include=../testports.properties
+osgi.console=7777

--- a/dev/com.ibm.ws.cdi.ejb_fat/publish/servers/cdi12MdbServer/server.xml
+++ b/dev/com.ibm.ws.cdi.ejb_fat/publish/servers/cdi12MdbServer/server.xml
@@ -1,0 +1,57 @@
+<server description="Server for testing MDBs with CDI">
+
+    <include location="../fatTestPorts.xml" />
+
+    <featureManager>
+        <feature>osgiconsole-1.0</feature>
+        <feature>cdi-1.2</feature>
+        <feature>ejbLite-3.2</feature>
+        <feature>mdb-3.2</feature>
+        <feature>servlet-3.1</feature>
+        <feature>componentTest-1.0</feature>
+        <feature>wasJmsServer-1.0</feature>
+        <feature>wasJmsClient-2.0</feature>
+        <feature>jms-2.0</feature>
+    </featureManager>
+
+    <!-- Settings for basicMdb.war -->
+
+    <jmsQueueConnectionFactory jndiName="jms/qcf">
+        <properties.wasJms />
+    </jmsQueueConnectionFactory>
+
+    <messagingEngine>
+        <queue id="basic/queue" />
+    </messagingEngine>
+
+    <jmsQueue id="jms/basic/queue" jndiName="jms/basic/queue">
+        <properties.wasJms queueName="basic/queue" />
+    </jmsQueue>
+    <jmsActivationSpec id="basicMdb/BasicMdb">
+        <properties.wasJms destinationRef="jms/basic/queue" />
+    </jmsActivationSpec>
+    
+    <!-- Settings for MdbTest.ear -->
+    
+    <messagingEngine>
+        <queue id="jar/jarQueue" />
+        <queue id="jar/jarNoDiscoveryQueue" />
+    </messagingEngine>
+
+    <jmsQueue id="jms/jar/jarQueue" jndiName="jms/jar/jarQueue">
+        <properties.wasJms queueName="jar/jarQueue" />
+    </jmsQueue>
+    <jmsActivationSpec id="MdbTest/EjbJar/JarMdb">
+        <properties.wasJms destinationRef="jms/jar/jarQueue" />
+    </jmsActivationSpec>
+    
+    <jmsQueue id="jms/jar/jarNoDiscoveryQueue" jndiName="jms/jar/jarNoDiscoveryQueue">
+        <properties.wasJms queueName="jar/jarNoDiscoveryQueue" />
+    </jmsQueue>
+    <jmsActivationSpec id="MdbTest/EjbNoDiscoveryJar/JarMdbNotDiscovered">
+        <properties.wasJms destinationRef="jms/jar/jarNoDiscoveryQueue" />
+    </jmsActivationSpec>
+    
+    <logging traceSpecification="JCDI=all"/>
+
+</server>

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/inject/InjectInjectionObjectFactory.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/inject/InjectInjectionObjectFactory.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -73,8 +73,8 @@ public class InjectInjectionObjectFactory {
 
     @SuppressWarnings("rawtypes")
     public static Object getObjectInstance(InjectInjectionBinding iBinding, Object targetObject, final CreationalContext<Object> cc,
-            final CreationalContext<Object> methodInvocactionContext,
-            CDIRuntime cdiRuntime) throws Exception {
+                                           final CreationalContext<Object> methodInvocactionContext,
+                                           CDIRuntime cdiRuntime) throws Exception {
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
             Tr.entry(tc, "getObjectInstance", new Object[] { iBinding, Util.identity(targetObject), Util.identity(cc), Util.identity(methodInvocactionContext) });
@@ -140,7 +140,9 @@ public class InjectInjectionObjectFactory {
                     EjbDescriptor<?> ejbDescriptor = ejbDescriptors.iterator().next();
                     String ejb_name = ejbDescriptor.getEjbName();
                     EjbDescriptor<?> realEjbDescriptor = weldManager.getEjbDescriptor(ejb_name);
-                    bean = weldManager.getBean(realEjbDescriptor);
+                    if (realEjbDescriptor != null) {
+                        bean = weldManager.getBean(realEjbDescriptor);
+                    }
                 }
 
             }
@@ -168,14 +170,13 @@ public class InjectInjectionObjectFactory {
                     }
                     Object ref = null;
                     if ((injectionPoint.getAnnotated() instanceof AnnotatedParameter<?>)
-                            && injectionPoint.getAnnotated().getBaseType().equals(InjectionPoint.class)){
+                        && injectionPoint.getAnnotated().getBaseType().equals(InjectionPoint.class)) {
                         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                             Tr.debug(tc, "Creating an InjectionPoint for {0}", injectionPoint);
                         }
                         ref = localBeanManager.createInjectionPoint((AnnotatedParameter<?>) injectionPoint.getAnnotated());
-                    }
-                    else if ((injectionPoint.getAnnotated() instanceof AnnotatedParameter<?>)
-                            && (((AnnotatedParameter<?>) injectionPoint.getAnnotated()).isAnnotationPresent(TransientReference.class))) {
+                    } else if ((injectionPoint.getAnnotated() instanceof AnnotatedParameter<?>)
+                               && (((AnnotatedParameter<?>) injectionPoint.getAnnotated()).isAnnotationPresent(TransientReference.class))) {
                         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                             Tr.debug(tc, "Getting an injectable ref from the bean manager with {0}, {1}", injectionPoint, Util.identity(methodInvocactionContext));
                         }
@@ -200,8 +201,8 @@ public class InjectInjectionObjectFactory {
                 ResolvableBuilder resolvableBuilder = new ResolvableBuilder(injectionPoint, beanManagerImpl);
                 Bean maybeProducerBean = beanManagerImpl.getBean(resolvableBuilder.create());
                 if (maybeProducerBean instanceof org.jboss.weld.bean.ProducerMethod
-                        && javax.enterprise.context.Dependent.class.isAssignableFrom(
-                                maybeProducerBean.getScope())) {
+                    && javax.enterprise.context.Dependent.class.isAssignableFrom(
+                                                                                 maybeProducerBean.getScope())) {
                     references.add(new InjectableNull());
                 }
 
@@ -210,7 +211,7 @@ public class InjectInjectionObjectFactory {
 
         debugInjectionObjects(references.toArray());
         if (references.size() == 1) {
-                Object reference = references.get(0);
+            Object reference = references.get(0);
             if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
                 Tr.exit(tc, "getObjectInstance", Util.identity(reference));
             }
@@ -220,11 +221,11 @@ public class InjectInjectionObjectFactory {
             if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
                 StringBuilder exitStringBuilder = new StringBuilder("[");
                 boolean first = true;
-                for (Object object : referencesArray){ 
-                    if(!first){
+                for (Object object : referencesArray) {
+                    if (!first) {
                         exitStringBuilder.append(", ");
-                    }
-                    else first = false;
+                    } else
+                        first = false;
                     exitStringBuilder.append(Util.identity(object));
                 }
                 exitStringBuilder.append("]");
@@ -235,7 +236,7 @@ public class InjectInjectionObjectFactory {
     }
 
     private static List<InjectionPoint> findUnmanagedInjectionPoints(BeanManager beanManager, WebSphereBeanDeploymentArchive bda, Class<?> targetClass,
-            Member targetMember) throws CDIException {
+                                                                     Member targetMember) throws CDIException {
         List<InjectionPoint> injectionPoints = new ArrayList<InjectionPoint>();
         List<InjectionPoint> allInjectionPoints = null;
         if (bda != null) {
@@ -354,4 +355,3 @@ public class InjectInjectionObjectFactory {
         }
     }
 }
-

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIEJBManagedObjectFactoryImpl.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIEJBManagedObjectFactoryImpl.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -12,9 +12,9 @@
  *******************************************************************************/
 package com.ibm.ws.cdi.impl.managedobject;
 
+import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
-import java.util.Iterator;
 import java.util.Set;
 
 import javax.enterprise.inject.spi.Bean;
@@ -96,7 +96,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
     @Override
     protected synchronized Bean<T> getBean() {
         if (!ejbBeanLookupComplete) {
-           doInitialization();
+            doInitialization();
         }
         return this.ejbBean;
     }
@@ -107,7 +107,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
 
             //This is needed for an edge case where the EJB is annotated @Vetoed.
             //In this case we'll find a ejbDescriptor but it will not have an associated bean.
-            //And we need to set these three variables. 
+            //And we need to set these three variables.
             EjbDescriptor<T> firstDescriptor = null;
             WeldManager firstBeanManager = null;
             WebSphereBeanDeploymentArchive firstBDA = null;
@@ -126,7 +126,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                 beanManager = (WeldManager) getCDIRuntime().getCurrentBeanManager();
             }
 
-            EjbDescriptor<T> ejbDescriptor = beanManager.getEjbDescriptor(this.ejbName);
+            EjbDescriptor<T> ejbDescriptor = lookUpEjbDescriptor(this.ejbName, bda, beanManager);
             if (ejbDescriptor != null) {
                 if (firstDescriptor == null) {
                     firstDescriptor = ejbDescriptor;
@@ -136,7 +136,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                 bean = beanManager.getBean(ejbDescriptor);
             }
             //If we didn't find a bean or found the wrong bean from the EJB descriptor (possible if multiple EJBs have the smae class name) continue looking.
-            if (bean == null || ! bean.getBeanClass().equals(getManagedObjectClass())) {
+            if (bean == null || !bean.getBeanClass().equals(getManagedObjectClass())) {
                 ejbDescriptor = null;
                 bean = null;
                 Set<WebSphereBeanDeploymentArchive> children = bda.getDescendantBdas();
@@ -144,7 +144,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                 while (bean == null && itr.hasNext()) {
                     bda = itr.next();
                     beanManager = (WeldManager) bda.getBeanManager();
-                    ejbDescriptor = beanManager.getEjbDescriptor(this.ejbName);
+                    ejbDescriptor = lookUpEjbDescriptor(this.ejbName, bda, beanManager);
                     if (ejbDescriptor != null) {
                         if (firstDescriptor == null) {
                             firstDescriptor = ejbDescriptor;
@@ -153,7 +153,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                         }
                         bean = beanManager.getBean(ejbDescriptor);
                     }
-                    if (bean != null && ! bean.getBeanClass().equals(getManagedObjectClass())) {
+                    if (bean != null && !bean.getBeanClass().equals(getManagedObjectClass())) {
                         partialMatchDescriptors.add(ejbDescriptor);
                         ejbDescriptor = null;
                         bean = null;
@@ -167,7 +167,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                     while (bean == null && itr.hasNext()) {
                         bda = itr.next();
                         beanManager = (WeldManager) bda.getBeanManager();
-                        ejbDescriptor = beanManager.getEjbDescriptor(this.ejbName);
+                        ejbDescriptor = lookUpEjbDescriptor(ejbName, bda, beanManager);
                         if (ejbDescriptor != null) {
                             if (firstDescriptor == null) {
                                 firstDescriptor = ejbDescriptor;
@@ -176,7 +176,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                             }
                             bean = beanManager.getBean(ejbDescriptor);
                         }
-                        if (bean != null && ! bean.getBeanClass().equals(getManagedObjectClass())) {
+                        if (bean != null && !bean.getBeanClass().equals(getManagedObjectClass())) {
                             partialMatchDescriptors.add(ejbDescriptor);
                             ejbDescriptor = null;
                             bean = null;
@@ -194,15 +194,16 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                     if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                         Tr.debug(tc, "Setting ejbDescriptor to firstEjbDescriptor: " + firstDescriptor.toString());
                     }
-                 } else {
+                } else {
                     Tr.error(tc, "Could not find an EjbDescriptor for : " + this.ejbName);
-                 }
+                }
             }
 
             if (bean == null) {
-                //The bean can be null if it's Vetoed. 
+                //The bean can be null if it's Vetoed.
                 if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                    Tr.debug(tc, "Found the following EjbDescriptor " + firstDescriptor + " for ejbName " + this.ejbName + " but it did not produce a bean or produced the wrong bean. It will be used anyway as this is expected in some circumstances");
+                    Tr.debug(tc, "Found the following EjbDescriptor " + firstDescriptor + " for ejbName " + this.ejbName
+                                 + " but it did not produce a bean or produced the wrong bean. It will be used anyway as this is expected in some circumstances");
                 }
 
                 //Use the firstDescriptor to match the old behaviour. Since the bean is null it will have gone looking for other descriptors.
@@ -213,7 +214,7 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                 }
 
                 //Output the others we found in the trace, it might come in handy.
-                if (! partialMatchDescriptors.isEmpty()) {
+                if (!partialMatchDescriptors.isEmpty()) {
                     String listStr = "[";
                     for (EjbDescriptor<T> desc : partialMatchDescriptors) {
                         listStr += desc.getEjbName() + ",";
@@ -231,9 +232,9 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
                 if (!bean.getBeanClass().equals(getManagedObjectClass())) {
                     //this exception should never happen
                     throw new IllegalStateException("Managed Class {" + getManagedObjectClass().getName() + "} does not match Bean Class {"
-                            + bean.getBeanClass().getName() + "}");
-                } else { 
-                    //We have an ejbDescriptor and a bean. The checks have pased, success. 
+                                                    + bean.getBeanClass().getName() + "}");
+                } else {
+                    //We have an ejbDescriptor and a bean. The checks have pased, success.
                     if (ejbDescriptor != null) {
                         this.ejbDescriptor = ejbDescriptor;
                         this.ejbBeanManager = beanManager;
@@ -245,6 +246,20 @@ public class CDIEJBManagedObjectFactoryImpl<T> extends AbstractManagedObjectFact
 
             ejbBeanLookupComplete = true;
         }
+    }
+
+    private EjbDescriptor<T> lookUpEjbDescriptor(String ejbName, WebSphereBeanDeploymentArchive bda, WeldManager bm) {
+
+        EjbDescriptor<T> result = bm.getEjbDescriptor(ejbName);
+
+        if (result == null && bda != null) {
+            Set<EjbDescriptor<?>> descriptors = bda.getEjbDescriptor(getManagedObjectClass());
+            if (descriptors != null) {
+                result = (EjbDescriptor<T>) descriptors.stream().filter(e -> ejbName.equals(e.getEjbName())).findFirst().orElse(null);
+            }
+        }
+
+        return result;
     }
 
     private synchronized EjbDescriptor<T> getEjbDescriptor() {

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -17,6 +17,7 @@ import java.security.AccessController;
 import java.security.PrivilegedAction;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -298,12 +299,9 @@ public class BeanDeploymentArchiveImpl implements WebSphereBeanDeploymentArchive
         }
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            String debugClassesString = this.beanClasses.entrySet()
-            .stream()
-            .filter(Objects::nonNull)
-            .map(entry -> entry.getKey()  + " = " + entry.getValue().toString())
-            .collect(Collectors.joining(", "));
-            Tr.debug(tc, "scan [ " + getHumanReadableName() + " ] AFTER SCAN. Bean classes: { " + debugClassesString + "}" );
+            String debugClassesString = this.beanClasses.entrySet().stream().filter(Objects::nonNull).map(entry -> entry.getKey() + " = "
+                                                                                                                   + entry.getValue().toString()).collect(Collectors.joining(", "));
+            Tr.debug(tc, "scan [ " + getHumanReadableName() + " ] AFTER SCAN. Bean classes: { " + debugClassesString + "}");
         }
 
         this.hasBeans = this.beanClasses.size() > 0;
@@ -745,7 +743,16 @@ public class BeanDeploymentArchiveImpl implements WebSphereBeanDeploymentArchive
      */
     @Override
     public Collection<EjbDescriptor<?>> getEjbs() {
-        return ejbDescriptors;
+        if (beanDiscoveryMode == BeanDiscoveryMode.NONE) {
+            // Don't tell Weld about any EJBs if bean discovery mode is none, otherwise it will turn them into beans
+            // which we don't want.
+
+            // We still need to store the EJB Descriptors so that we can find which BDA contains a given EJB, even if
+            // it has bean discovery mode none.
+            return Collections.emptySet();
+        } else {
+            return ejbDescriptors;
+        }
     }
 
     private Set<Class<?>> getEJBClasses() {
@@ -871,17 +878,18 @@ public class BeanDeploymentArchiveImpl implements WebSphereBeanDeploymentArchive
     public void addEjbDescriptor(EjbDescriptor<?> ejbDescriptor) {
         Tr.entry(tc, "addEjbDescriptor. Adding EjbDescriptor: " + ejbDescriptor + " with ejbName: " + ejbDescriptor.getEjbName() + " to bda: " + getHumanReadableName());
 
+        this.ejbDescriptors.add(ejbDescriptor);
+        Class<?> beanClass = ejbDescriptor.getBeanClass();
+        Set<EjbDescriptor<?>> ejbDescriptors = ejbDescriptorMap.get(beanClass);
+        if (ejbDescriptors == null) {
+            ejbDescriptors = new HashSet<EjbDescriptor<?>>();
+        }
+        ejbDescriptors.add(ejbDescriptor);
+        this.ejbDescriptorMap.put(beanClass, ejbDescriptors);
+
         if (getBeanDiscoveryMode() != BeanDiscoveryMode.NONE) {
-            this.ejbDescriptors.add(ejbDescriptor);
-            Class<?> beanClass = ejbDescriptor.getBeanClass();
             this.beanClasses.put(beanClass.getName(), beanClass);
             this.ejbClasses.add(beanClass);
-            Set<EjbDescriptor<?>> ejbDescriptors = ejbDescriptorMap.get(beanClass);
-            if (ejbDescriptors == null) {
-                ejbDescriptors = new HashSet<EjbDescriptor<?>>();
-            }
-            ejbDescriptors.add(ejbDescriptor);
-            this.ejbDescriptorMap.put(beanClass, ejbDescriptors);
             this.hasBeans = true;
         }
     }

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/JndiHelperComponentMetaData.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/JndiHelperComponentMetaData.java
@@ -4,7 +4,7 @@
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-2.0/
- * 
+ *
  * SPDX-License-Identifier: EPL-2.0
  *
  * Contributors:
@@ -13,11 +13,12 @@
 package com.ibm.ws.cdi.liberty;
 
 import com.ibm.websphere.csi.J2EEName;
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.ws.container.service.metadata.extended.IdentifiableComponentMetaData;
 import com.ibm.ws.runtime.metadata.ApplicationMetaData;
 import com.ibm.ws.runtime.metadata.ComponentMetaData;
 import com.ibm.ws.runtime.metadata.MetaDataImpl;
 import com.ibm.ws.runtime.metadata.ModuleMetaData;
-import com.ibm.ws.container.service.metadata.extended.IdentifiableComponentMetaData;
 
 /**
  * A Component MetaData used for jndi validation
@@ -41,6 +42,7 @@ public class JndiHelperComponentMetaData extends MetaDataImpl implements Compone
     }
 
     @Override
+    @Trivial
     public J2EEName getJ2EEName() {
         return module.getJ2EEName();
     }

